lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -139,95 +150,275 @@
           }
         }
 '''
+import json
+
+from abc import ABCMeta, abstractmethod
+
+from ansible.module_utils._text import to_native, to_text
+from ansible.module_utils.basic import AnsibleModule, get_all_subclasses
+from ansible.module_utils.six import with_metaclass
+
+
+class TestFailed(Exception):
+    pass
+
+
+class PkgMgr(with_metaclass(ABCMeta, object)):
+
+    def __init__(self):
+
+        if not self.test():
+            raise TestFailed("Test failed to pass")
+
+    @abstractmethod
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        pass
+
+    @abstractmethod
+    def list_installed(self):
+        ''' This method should return a list of installed packages, each list item will be passed to get_package_details '''
+        pass
+
+    @abstractmethod
+    def get_package_details(self, package):
```

Comment:
It is better to have consistent method naming: sometimes you use "package" in names, sometimes you don't.
I suggest reconsidering it and maybe omit word "package" at all. In this case, for example, you use it as an argument, which makes it obvious even when you change it to `get_details_of(self, package)` what it's about. Otherwise you can choose to use "package" in all related methods, consistency matters.